Feature 348 dmap reformat (#354)
dtcenter/METdatadb: docs/Users_Guide/reformat_stat_data.rst

===========

The METreformat module provides support for reformatting/rearranging MET .stat/.tcst output into METplotpy-readable

> input formats.  The formats vary, based on the plot type (e.g. a histogram plot requires ranks and values column data,

Reviewer: ```suggestion
input formats.  The formats vary, based on the plot type.  For example, a histogram plot requires ranks and values column data,
```

> whereas a line plot will need statistic name and statistic value columns). The MET .stat/.tcst ASCII output generated

Reviewer: ```suggestion
whereas a line plot will need statistic name and statistic value columns. The MET .stat/.tcst ASCII output generated
```

> by MET tools such as the point-stat, grid-stat, ensemble-stat, and tc-pairs tools may contain ASCII columnar data

Reviewer: ```suggestion
by MET tools such as the Point-Stat, Grid-Stat, Ensemble-Stat, and TC-Pairs tools may contain ASCII columnar data
```

> from multiple line types (determined by settings in the MET configuration file).

Reviewer: ```suggestion
from multiple line types, as determined by settings in the MET configuration file.
```

.. note ::
  Currently, there are MET linetypes that **do not** have reformatter support.  When an unsupported linetype
  is requested, an appropriate error message will be generated.

MET .stat and .tcst files can be comprised of numerous rows of data with different number of columns. This is a

> result of requesting multiple linetypes in a MET configuration file (e.g. requesting .

Reviewer: @bikegeek This line and the next line currently read:
"This is a result of requesting multiple linetypes in a MET configuration file (e.g. requesting . All MET .stat line types have common columns".  Is that how it should read?  I'm confused by the ". All" part and am wondering if it should be "This is a result of requesting multiple linetypes in a MET configuration file (e.g. requesting all MET .stat line types have common columns" instead?

> All MET .stat line types have common columns (refer to table 11.1 in the MET User's Guide) that are labelled

Reviewer: ```suggestion
All MET .stat line types have common columns (refer to `Table 11.1 <https://met.readthedocs.io/en/main_v12.0/Users_Guide/point-stat.html#id5>`_ in the MET User's Guide) that are labelled
```

(i.e. the columns have headers/names).  The remaining columns

> are unlabelled.  When the stat output from the MET point-stat, grid-stat, ensemble-stat, or tc-pairs tools are

Reviewer: ```suggestion
are unlabelled.  When the stat output from the MET Point-Stat, Grid-Stat, Ensemble-Stat, or TC-Pairs tools are
```

reformatted, these unlabelled columns are rearranged into a format appropriate

> for the METplotpy plot of interest (based on the *line_type* setting in the YAML configuration file).

Reviewer: ```suggestion
for the METplotpy plot of interest, based on the *line_type* setting in the YAML configuration file.
```

The format of input data used by the METplotpy plots is influenced by METviewer.
When generating plots in METviewer, the MET .stat input data is first loaded into the

> a database. A database query is then performed to filter the data (based on criteria specified through

Reviewer: ```suggestion
a database. A database query is then performed to filter the data, based on criteria specified through
```

> METviewer's graphical user interface).  The query is based on:

Reviewer: ```suggestion
METviewer's graphical user interface.  The query is based on:
```

- the selected values for the dependent variables (i.e. forecast variable) and the associated statistics of interest
- series variables (i.e. model and specific model values)
- fixed values (corresponding to columns in the MET output such as forecast lead, vx mask, model, etc.)

Any requested aggregation statistics (i.e. mean, median) are calculated by invoking the METcalcpy agg_stat.py code.
The format of the data from the query result may vary, based on the line type and whether aggregation statistics were
calculated.

.. note::

  METviewer invokes the appropriate METplotpy plot script using the database query results as input, and **it
  is this format that is expected by the METplotpy scripts**.

Users have the option to generate plots from the command line, by-passing METviewer and its database. **However,

> the MET stat-analysis** tool will need to be used for performing any necessary filtering of data (e.g. by any combination of

Reviewer: ```suggestion
the MET Stat-Analysis** tool will need to be used for performing any necessary filtering of data (e.g. by any combination of
```

> grid-stat, ensemble-stat, or tc-pairs tool must be reformatted before invoking the METplotpy scripts from the

Reviewer: ```suggestion
Grid-Stat, Ensemble-Stat, or TC-Pairs tools must be reformatted before invoking the METplotpy scripts from the
```

command line.

..  dropdown:: Linetypes, MET tools and corresponding plot types (click to expand)

     .. dropdown:: FHO

>           **Forecast, Hit, Observation Rates**

Reviewer: ```suggestion
         **Forecast, Hit, Observation Rates**
```
Alignment is off.  Attempting to fix.

>           From MET tools:

Reviewer: ```suggestion
         From MET tools:
```
Alignment is off. Attempting to fix.

>             - point-stat

Reviewer: ```suggestion
            - point-stat
```
These references to the MET tools from here and below should either be capitalized if referring to the MET tool (e.g. Point-Stat) or with all lowercase and with an underscore (e.g. point_stat) if referring to the executable.  Please fix up the references as intended and I'll take another look.  Thanks!

            - grid-stat

         Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: CNT

>          **Continuous Statistics**

Reviewer: ```suggestion
        **Continuous Statistics**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

        Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: CTC

>           **Contingency Table Counts**

Reviewer: ```suggestion
         **Contingency Table Counts**
```
Alignment is off. Attempting to fix.

>           From MET tools:

Reviewer: ```suggestion
         From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

         Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: CTS

>          **Contingency Table Statistics**

Reviewer: ```suggestion
        **Contingency Table Statistics**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

        Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: SL1L2

>          **Scalar L1L2 Partial Sums**

Reviewer: ```suggestion
        **Scalar L1L2 Partial Sums**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

        Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: VL1L2

>          **Vector L1L2 Partial Sums**

Reviewer: ```suggestion
        **Vector L1L2 Partial Sums**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

        Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: ECNT

>          **Ensemble Continuous Statistics**

Reviewer: ```suggestion
        **Ensemble Continuous Statistics**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - ensemble-stat

        Used by the following METplotpy plot types:
            - line (for ensemble spread-skill plots)

     .. dropdown:: MCTS

>          **Multi-category Contingency Table Statistics**

Reviewer: ```suggestion
        **Multi-category Contingency Table Statistics**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat

        Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: VCNT

>          **Vector Continuous Statistics**

Reviewer: ```suggestion
        **Vector Continuous Statistics**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

        Used by the following METplotpy plot types:
            - line
            - bar
            - box
            - contour
            - performance diagram
            - revision box
            - revision series
            - Taylor Diagram

     .. dropdown:: PCT

>          **Contingency Table counts for Probabilistic forecasts**

Reviewer: ```suggestion
        **Contingency Table counts for Probabilistic forecasts**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat
            - ensemble_stat

        Used by the following METplotpy plot types:
            - ROC diagram (Receiver Operator Curve)

     .. dropdown::  RHIST

>          **Ranked Histogram Counts**

Reviewer: ```suggestion
        **Ranked Histogram Counts**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - ensemble-stat

        Used by the following METplotpy plot types:
            - rank histogram (rhist)

     .. dropdown::  TCDIAG (with TCMPR)

>          **Tropcial Cyclone Diagnostics**

Reviewer: ```suggestion
        **Tropcial Cyclone Diagnostics**
```
Alignment is off. Attempting to fix.

>          **Tropical Cyclone Matched Pairs**

Reviewer: ```suggestion
        **Tropical Cyclone Matched Pairs**
```
Alignment is off. Attempting to fix.

Reviewer: **Tropcial Cyclone Diagnostics** and **Tropical Cyclone Matched Pairs** are on the same line when this is rendered in the documentation here https://metdataio.readthedocs.io/en/feature_348_dmap_reformat/Users_Guide/reformat_stat_data.html.  You may need to have a [line block](https://metplus.readthedocs.io/en/develop/Contributors_Guide/documentation.html#line-blocks).  If it is still problematic, I'm sure @lisagoodrich can help.

>          From MET tool:

Reviewer: ```suggestion
        From MET tool:
```
Alignment is off. Attempting to fix.

            - tc-pairs

        Used by the following METplotpy plot types:
            - TCMPR (Tropical Cyclone Matched Pair) plots:

> 

Reviewer: This extra line does not show up in RTD:
https://metdataio.readthedocs.io/en/feature_348_dmap_reformat/Users_Guide/reformat_stat_data.html
Please ask @lisagoodrich for assistance if needed.

            - mean line plot
            - median line plot
            - boxplot
            - relative performance plot
            - rank plot
            - mean skill line plot
            - median skill line plot

     .. dropdown:: MPR

>          **Matched Pair data**

Reviewer: ```suggestion
        **Matched Pair data**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

            - point-stat
            - grid-stat

        Used by the following METplotpy plot types:
            - line
            - scatter

     .. dropdown:: DMAP

>          **Distance Map Statistics**

Reviewer: ```suggestion
        **Distance Map Statistics**
```
Alignment is off. Attempting to fix.

>          From MET tools:

Reviewer: ```suggestion
        From MET tools:
```
Alignment is off. Attempting to fix.

plot and line type.  Formats fall into the following categories:

    .. dropdown::  by statistic name, statistic value and confidence values

       - Needed by the following plot types in METplotpy:

         - line
         - bar
         - box
         - contour
         - performance diagram
         - revision box
         - revision series
         - Taylor diagram

       - Unlabelled columns in the original .stat file are placed under statistic name, statistic value,
         and confidence level columns (in addition to the common stat columns):

         - stat_name
         - stat_value
         - stat_ncl

            - lower level normal confidence limit

         - stat_ncu

            - upper level normal confidence limit

         - stat_bcl

            - lower level bootstrap confidence limit

         - stat_bcu

            - upper level bootstrap confidence limit

       - The original format is converted from wide format (where each row corresponds to a unique
         set of common stat column values) to longer format (where there are multiple rows
         corresponding to the same common stat column values).

>            - The common stat columns are common to all line types for point-stat, grid-stat, and ensemble-stat (model, vx mask, description, fcst lead, etc.)

Reviewer: ```suggestion
           - The common stat columns are common to all line types for Point-Stat, Grid-Stat, and Ensemble-Stat (model, vx mask, description, fcst lead, etc.)
```

             as described in the MET User's Guide:

>               - table 11.1 (point-stat)

Reviewer: ```suggestion
              - `Table 11.1 <https://met.readthedocs.io/en/main_v12.0/Users_Guide/point-stat.html#id5>`_ (Point-Stat)
```

>               - table 12.1 (grid-stat)

Reviewer: ```suggestion
              - `Table 12.1 <https://met.readthedocs.io/en/main_v12.0/Users_Guide/grid-stat.html#id8>`_ (grid-stat)
```

>               - table 13.1 (ensemble-stat)

Reviewer: ```suggestion
              - `Table 13.1 <https://met.readthedocs.io/en/main_v12.0/Users_Guide/ensemble-stat.html#id1>`_ (ensemble-stat)
```

Author: modified and also used the camel-case version of the tool name, separated by '-'

         .. dropdown:: Unformatted Example:

            .. literalinclude:: ./figure/grid_stat_FV3_core_lsm1_020000L_20190521_020000V.stat

         - In the unformatted example:

           - each row represents unique data (tidy data)
           - the number of columns is different for each row, due to each row representing a different linetype
             (e.g. DMAP, NBRTC)
           - numerous columns are unlabelled

         .. dropdown:: Reformatted Example (truncated):

            ..  literalinclude:: ./figure/reformatted_dmap_for_lineplot_output.txt

         -  In the reformatted example:

            - reformatted data file contains only one linetype (in this example DMAP), as requested in the configuration file

            - multiple rows have duplicate common stat column values, resulting in more rows

>               - note that for the rows containing the first **fy** and **oy** stat_names, the common stat columns of

Reviewer: ```suggestion
              - note that for the rows containing the first **FY** and **OY** stat_names, the common stat columns of
```
Changing these to upper case to be consistent with the MET documentation (https://met.readthedocs.io/en/latest/Users_Guide/grid-stat.html#id9)

                version,..., linetype, and total are the same.  The unique data has been replicated and separated
                based on stat_name and stat_value.

    .. dropdown::  by all common stat columns AND linetype-specific columns from MET

       - Used by the following plot types and linetypes in METplotpy:

         - **scatter** plot

           - **MPR** linetype
           - **DMAP** linetype

         - unlabelled columns in the original .stat file are placed under the linetype-specific column names

>            - refer to the MET User's Guide for point-stat, grid-stat, ensemble-stat, and tc-pairs  for the column names

Reviewer: ```suggestion
           - refer to the MET User's Guide for Point-Stat, Grid-Stat, Ensemble-Stat, and TC-Pairs  for the column names
```

Author: suggested changes made during an edit of the file (outside of this interface)

       - The original format is converted from wide format (where each row corresponds to a unique
         set of common stat column values) to longer format (where there are multiple rows
         corresponding to the same common stat column values).

           - The common stat columns are common to all line types for point-stat, grid-stat, and ensemble-stat (model, vx mask, description, fcst lead, etc.)
             as described in the MET User's Guide:

>               - table 11.1 (point-stat)

Reviewer: ```suggestion
              - `Table 11.1 <https://met.readthedocs.io/en/main_v12.0/Users_Guide/point-stat.html#id5>`_ (Point-Stat)
```

Author: suggested change made while editing the file for the other tables.
